Show breakdown of file type in Contribution Bar

apoorva17 opened the pull request:
Fixes #456 

```
Contribution Bar is currently a single red bar which represents total contribution of 
the author.

Let's reflect the breakdown of filetype in Contribution Bar. 

The user can hover over a specific section of the Contribution Bar to see information 
about the filetype, total lines contributed and analysis period in the tooltip. In order 
to deal with the 'too many filetypes' scenario, the Contribution Bar only 
shows the breakdown for the first 5 filetypes.

Since this filetype information is currently being processed in the front-end in code 
view (v_authorship), let's move the processing of filetype information to the backend 
in order to be able to use the pre-processed filetype information for both chart 
view (v_summary) and code view (v_authorship). Let's also refactor the filetype
processing in code-view (v_authorship) to match the changes being made.
```

apoorva17 commented:
@reposense/devs ready for review

yong24s commented:
Looks really nice 🎉 

Quick comments:

May need to consider colors with higher contrast. 
Its quite difficult to spot Aditya and Jun An JavaScript contributions as Magenta is eclipsed by Red

![image](https://user-images.githubusercontent.com/2003406/51459244-d6577f80-1d92-11e9-8b19-ab0a5e560973.png)

@damithc 
Maybe we can extend the colors to the check boxes in code view in another PR? Or maybe its too colorful 😆

damithc commented:
I wonder if we should use the same color for all ramps, as we are now using colors for something else.

> Maybe we can extend the colors to the check boxes in code view in another PR? Or maybe its too colorful

hmm.. the contribution bar has only 5 types, there can be any number of tick boxes.Does it add value to the user?  :-)

eugenepeh commented:
>I wonder if we should use the same color for all ramps, as we are now using colors for something else.

it will be hard to differentiate overlapping ramps though.

apoorva17 commented:
> May need to consider colors with higher contrast.
> Its quite difficult to spot Aditya and Jun An JavaScript contributions as Magenta is eclipsed by Red
> 
> ![image](https://user-images.githubusercontent.com/2003406/51459244-d6577f80-1d92-11e9-8b19-ab0a5e560973.png)
> 
I am already selecting from 20 contrasting colors based on the color wheel, but since I am allocating the colors based on filetype, it's hard to ensure that adjacent bars are also contrasting colors.

apoorva17 commented:
> I wonder if we should use the same color for all ramps, as we are now using colors for something else.

Currently, ramps of same color represent commits made on same day, won't using the same color for all ramps lose that information?

damithc commented:
> Currently, ramps of same color represent commits made in the same time period by different authors, won't using the same color for all ramps lose that information?

I don't think that feature is very useful; there is not much value in identifying commits done in the same period. In any case, commits on the same period should be vertically aligned, which helps to identify them to some extent.

apoorva17 commented:
@eugenepeh i've updated to master and updated the tooltip, for your review

eugenepeh commented on src/main/java/reposense/authorship/model/AuthorshipSummary.java:
```diff
@@ -27,10 +31,29 @@ public void addAuthorContributionCount(Author author) {
+        TreeMap<String, Integer> fileTypeContributionMap = authorFileTypeContributionMap.get(author);
+
+        if (fileTypeContributionMap.containsKey(fileType)) {
+            fileTypeContributionMap.put(fileType, fileTypeContributionMap.get(fileType) +  1);
```
perhaps can remove this conditional to use `.getOrDefault`?

eugenepeh commented on src/main/java/reposense/authorship/model/AuthorshipSummary.java:
```diff
@@ -27,10 +31,29 @@ public void addAuthorContributionCount(Author author) {
+            fileTypeContributionMap.put(fileType, 1);
+        }
+
+        authorFileTypeContributionMap.put(author, fileTypeContributionMap);
```
is it necessary to bind back to the author if `authorFileTypeContributionMap.get(author);` works in the first place?

apoorva17 commented:
@eugenepeh how do i add information about file formats specified by user for each repo to ```summary.json```? Can't find where the fields for each repo has been specified since ```SummaryReportJson``` directly does ```this.repos = repos```. I need this information to display the formats specified by user first in Contribution Bar and legend.

eugenepeh commented:
Looks good! Can we hide the irrelevant file types in the legend?
Also, we probably need to think how to deal with those really small contribution bar but that can be another PR.

![image](https://user-images.githubusercontent.com/19277206/51784289-53169f00-2181-11e9-8bac-1d64f855816c.png)

apoorva17 commented:
> Looks good! Can we hide the irrelevant file types in the legend?
I have discussed the legend with the @damithc. We agreed upon showing the file types in order specified by user if user adds the file formats, or else show all the default file types in the legend for the repository.

apoorva17 commented:
@eugenepeh @ongspxm kindly review updated implementation

eugenepeh commented on frontend/src/static/css/style.scss:
```diff
@@ -149,11 +149,34 @@ header{
+                margin-top:0.25rem;
+                margin-left:0.25rem;
+                margin-right:0.25rem;
+
```
unnecessary newline

eugenepeh commented on frontend/src/static/css/style.scss:
```diff
@@ -149,11 +149,34 @@ header{
+
+    &__filetype--breakdown{
+        &__title{
+
```
same for here

eugenepeh commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -193,7 +181,7 @@ window.vAuthorship = {
     getFileBlankLineInfo(fileType) {
       return fileType + ': ' + 'Blank: ' + this.filesBlankLinesObj[fileType]
-          + ', Non-Blank: ' + (this.filesLinesObj[fileType] - this.filesBlankLinesObj[fileType]);
+          + ', Non-Blank: ' + (this.info.filesLinesObj[fileType] - this.filesBlankLinesObj[fileType]);
```
why is it that `filesLinesObj` and `filesBlankLinesObj` are now stored in different level of abstraction?

apoorva17 replied:
```filesLinesObj``` is now processed in the backend, therefore it is part of ```info``` prop now. However, all blank lines related information is still processed in the frontend, therefore ```filesBlankLinesObj``` is not part of the ```info``` prop.

apoorva17 commented on src/systemtest/resources/dateRange/expected/reposense_testrepo-Beta_master/authorship.json:
```diff
@@ -1,6 +1,7 @@
 [
   {
     "path": "src/main/resources/view/MainWindow.fxml",
+    "fileType": "fxml",
```
I am currently processing filetype related information in the backend, therefore i need information about the filetype of each file to populate the filetype based contribution for each author. I also feel that filetype of each file is an important information about the file which can easily be extracted from the JSON for any future work involving the filetypes.

eugenepeh replied:
Yes, but with this change, comes the responsibility of fixing everything that are relevant. 
If you're not going to make use of it in the front, I would suggest making it a `fileType` a transient variable and open an issue to add it into the json and fix the relevant stuff (e.g. select file type) together afterward.

apoorva17 replied:
Ok, i've removed ```fileType``` from ```File Result``` and the json file.

ongspxm commented on frontend/src/index.jade:
```diff
@@ -148,10 +158,10 @@ html
                 span.loc(v-bind:title="getTotalFileBlankLineInfo()") {{ totalLineCount }}&nbsp;
                 span.bloc(v-bind:title="getTotalFileBlankLineInfo()") ({{ totalBlankLineCount }})
-              template(v-for="type in Object.keys(filesLinesObj)")
+              template(v-for="type in Object.keys(info.filesLinesObj)", v-if="info.filesLinesObj[type] > 0")
```
dun use `v-if` and `v-for` in the same object, and do `v-bind:key` for each `v-for` obj.

do `template(v-if='...')`

apoorva17 replied:
I've filtered ```filesLinesObj``` by iterating over a computed property ```filesExistingLinesObj``` instead as per this: https://vuejs.org/v2/style-guide/#Avoid-v-if-with-v-for-essential

apoorva17 commented:
@damithc ramp colors have been reset to original colors as discussed

eugenepeh commented on frontend/src/index.jade:
```diff
@@ -132,13 +141,15 @@ html
                 .summary-chart__contrib--bar(
-                  v-for="width in getContributionBars(user.totalCommits)",
-                  v-bind:style="{ width: width+'%' }"
+                  v-for="width in widths",
```
here should be double indented

eugenepeh commented on frontend/src/index.jade:
```diff
@@ -132,13 +141,15 @@ html
-                  v-bind:style="{ width: width+'%' }"
+                  v-for="width in widths",
+                  v-bind:style="{ width: width + '%', 'background-color'\
+                      : contributionBarColors[user.repoPath][fileType]}",
```
perhaps use a higher level break, at the comma, joining the `background-color` to the next line?

eugenepeh commented on frontend/src/index.jade:
```diff
@@ -132,13 +141,15 @@ html
+                  v-bind:style="{ width: width + '%', 'background-color'\
+                      : contributionBarColors[user.repoPath][fileType]}",
+                  v-bind:title= "fileType + ': ' + user.fileTypeContribution[fileType] + ' lines, total: '\
+                      + user.totalCommits + ' lines ' + '(contribution from ' + minDate + ' to ' + maxDate + ')'"
```
similarly here

apoorva17 replied:
I assume u are referring to both indentation and joining ```total

eugenepeh commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -111,12 +118,12 @@ window.vAuthorship = {
           totalBlankLineCount += segmentInfo.blankLineCount;
-          this.addLineCountToFileType(file.path, segmentInfo.blankLineCount, filesBlanksInfoObj);
+          this.addBlankLineCountToFileType(file.path, segmentInfo.blankLineCount,
+            filesBlanksInfoObj);
```
double indentation
@ongspxm seems like this is not covered by eslint, can you open a PR to fix it?

apoorva17 replied:
@eugenepeh CI is failing due to this double indentation which you asked me to add

eugenepeh commented on frontend/src/static/js/v_summary.js:
```diff
@@ -145,21 +146,39 @@ window.vSummary = {
-      }
+        const cnt = parseInt(contribution / contributionLimit, 10);
+        for (let cntId = 0; cntId < cnt; cntId += 1) {
+          res.push(100);
```
seems like 100 denote the max length of contribution bar. Perhaps make it a constant?

eugenepeh commented on frontend/src/static/js/v_summary.js:
```diff
@@ -282,6 +301,32 @@ window.vSummary = {
+        const colors = {};
+        let i = 0;
+        if (Object.keys(user.fileTypeContribution).length === 15) {
+          user.fileTypeContribution = this.sortKeysAlphabetically(user.fileTypeContribution);
```
why is it only sorted if its === 15?

apoorva17 replied:
It was meant to check if the file formats are the default file formats, as it should only be sorted alphabetically if they are the default file formats, otherwise ordering should remain the same. I have replaced this with a ```isDefaultFileFormats``` method for better checking

eugenepeh commented on frontend/src/static/js/v_summary.js:
```diff
@@ -282,6 +301,32 @@ window.vSummary = {
+        }
+        Object.keys(user.fileTypeContribution).forEach((fileType) => {
+          colors[fileType] = selectedColors[i];
+          i = (i + 1) % 20;
```
if the 20 represents the size of the array, use the size directly.

eugenepeh commented on src/main/java/reposense/authorship/FileResultAggregator.java:
```diff
@@ -15,15 +16,17 @@
-    public static AuthorshipSummary aggregateFileResult(List<FileResult> fileResults, List<Author> authors) {
-        AuthorshipSummary authorContributionSummary = new AuthorshipSummary(fileResults, authors);
+    public static AuthorshipSummary aggregateFileResult(List<FileResult> fileResults, List<Author> authors,
+                                                            List<Format> formats) {
```
indentation

eugenepeh commented on src/main/java/reposense/authorship/FileResultAggregator.java:
```diff
@@ -15,15 +16,17 @@
                     continue;
                 }
                 authorContributionSummary.addAuthorContributionCount(author);
+                authorContributionSummary.addAuthorFileTypeContributionCount(author, fileResult.getFileType());
```
cant this be together with `addAuthorContributionCount`?

apoorva17 replied:
Ok, i've removed the separate ```addAuthorFileTypeContributionCount``` method and combined it with ```addAuthorContributionCount```

eugenepeh commented on src/main/java/reposense/authorship/model/FileResult.java:
```diff
@@ -11,11 +11,14 @@
     public FileResult(String path, ArrayList<LineInfo> lines, HashMap<Author, Integer> authorContributionMap) {
         this.path = path;
+        String fileType = path.substring(path.lastIndexOf('.') + 1);
+        this.fileType = (fileType.length() == 0) ? "others" : fileType;
```
why the length access when you're checking if its empty?

apoorva17 commented:
@eugenepeh requested changes have been made

eugenepeh commented on frontend/src/static/js/v_summary.js:
```diff
@@ -282,6 +302,43 @@ window.vSummary = {
+        const colors = {};
+        let i = 0;
+        if (this.isDefaultFileFormats(Object.keys(user.fileTypeContribution))) {
+          user.fileTypeContribution = this.sortKeysAlphabetically(user.fileTypeContribution);
```
What I actually meant is, why don't we just sort it regardless of whether is it default file formats?
Making a const `defaultFileFormats` is not really useful because we will have to keep PR which modifies the default file formats in check so that the devs who change the java file will remember to change for the js too.

eugenepeh replied:
By right, there shouldn't be a need to sort the default file formats in the first place as the json would outputting the java objects in the same order every time. We would have problem passing the test cases if it didn't.

apoorva17 replied:
> What I actually meant is, why don't we just sort it regardless of whether is it default file formats?

This was actually my original implementation, which was modified after my discussion with @damithc. We agreed that if the user specifies the file formats, then the contribution bar and legend should reflect those file formats in order specified by the user, and not alphabetically. If no file format is specified by the user, then all default file formats should be shown in the legend in alphabetical order. Since information about the default file formats is not passed from the backend to frontend, i need to create a const ```defaultFileFormats``` for the sorting.

apoorva17 replied:
> By right, there shouldn't be a need to sort the default file formats in the first place as the json would outputting the java objects in the same order every time. We would have problem passing the test cases if it didn't.

What do you mean? I am using a ```HashMap``` in the backend which doesn't maintain any ordering of the keys, it just adds/increments file type count as encountered in ```FileResultAggregator```. Since the ordering is alphabetical only for ```defaultFileFormat```, there is a need to sort it in the frontend.

eugenepeh replied:
>What do you mean? I am using a `HashMap` in the backend which doesn't maintain any ordering of the keys, it just adds/increments file type count as encountered in `FileResultAggregator`. Since the ordering is alphabetical only for `defaultFileFormat`, there is a need to sort it in the frontend.

In that case, how would you be sure that the legend will always `reflect those file formats in order specified by the user`?

apoorva17 replied:
I have changed the implementation to use a ```LinkedHashMap``` which uses the ordering of ```DEFAULT_FORMATS``` specified in the backend which is already sorted alphabetically, and uses the file formats in order specified by user, therefore I've removed the sorting in the frontend.

eugenepeh commented on frontend/src/index.jade:
```diff
@@ -132,13 +141,16 @@ html
+                        'background-color': contributionBarColors[user.repoPath][fileType] }",
+                    v-bind:title="fileType + ': ' + user.fileTypeContribution[fileType] + ' lines, '\
+                        + 'total: ' + user.totalCommits + ' lines ' + '(contribution from ' + minDate + ' to '\
+                            + maxDate + ')'"
```
missed out this one. if this is a continuing line wrap, dont need to further indent.

eugenepeh commented on src/main/java/reposense/authorship/FileResultAggregator.java:
```diff
@@ -15,15 +16,16 @@
-    public static AuthorshipSummary aggregateFileResult(List<FileResult> fileResults, List<Author> authors) {
-        AuthorshipSummary authorContributionSummary = new AuthorshipSummary(fileResults, authors);
+    public static AuthorshipSummary aggregateFileResult(List<FileResult> fileResults, List<Author> authors,
+                                                        List<Format> formats) {
```
indentation

apoorva17 commented:
@eugenepeh requested changes have been made

eugenepeh commented:
Please ensure that the CI pass before requesting for reviews, and also take note of my previous comment

>For non-trivial pr, avoid rebasing so that we can see the difference from last iteration.

apoorva17 commented:
> Please ensure that the CI pass before requesting for reviews, 

CI is failing because of a double indentation you asked me to add as I have highlighted above, everything else is passing

eugenepeh commented:
>CI is failing because of a double indentation you asked me to add as I have highlighted above, everything else is passing

I see, can you open a PR to quick fix this? https://eslint.org/docs/rules/indent#functiondeclaration

apoorva17 commented:
> I see, can you open a PR to quick fix this? https://eslint.org/docs/rules/indent#functiondeclaration

A separate issue should be created as this seems to be related to #502 , not my PR.

eugenepeh commented:
>A separate issue should be created as this seems to be related to #502 , not my PR.

It is a courtesy to fix things that are found broken when doing your own work; software engineering is never about individualism, its a collaborative effort. It is also why it is important to provide critical reviews; if you don't want to fix problems created by others, make sure they do it right in the first place.

apoorva17 commented:
> It is a courtesy to fix things that are found broken when doing your own work; software engineering is never about individualism, its a collaborative effort. It is also why it is important to provide critical reviews; if you don't want to fix problems created by others, make sure they do it right in the first place.

I would have, but I am not familiar with how ESLint works.

eugenepeh commented:
I believe it is just adding the rule into https://github.com/reposense/RepoSense/blob/master/.eslintrc.json.

Since this is a quick simple fix, we can omit proposed message for it.

apoorva17 commented:
@eugenepeh i've created the PR for the quick fix

eugenepeh commented:
>@eugenepeh i've created the PR for the quick fix

Thanks! It has been merged, do proceed to update your PR.

apoorva17 commented:
@eugenepeh adding the rule still seems to be giving the same error

eugenepeh commented:
> @eugenepeh adding the rule still seems to be giving the same error

Did you check if it works before submitting the other PR?

apoorva17 commented:
> I see, can you open a PR to quick fix this? https://eslint.org/docs/rules/indent#functiondeclaration

I just used inserted the rule from the link you had put in your previous comment

eugenepeh commented:
>I just used inserted the rule from the link you had put in your previous comment

The main purpose is of course to fix the error. Please test it out and submit a PR again.

apoorva17 commented:
> The main purpose is of course to fix the error. Please test it out and submit a PR again.

Hmm, i tried various rules but it doesn't seem to solve the error. Maybe it shouldn't be double indentation for that case? If you still think it is double indentation, please point out the rule and parameters for the rule you are referring to.

yamidark commented:
@apoorva17 that line is making a function call, so the rule that we need to add is [CallExpression](https://eslint.org/docs/rules/indent#callexpression). 
Note that some of the lines in `main.js` will be violated when you add this in, since we forgot to check and add this rule and went to convert them, so do help in correcting those as well.

apoorva17 commented:
@eugenepeh sorry for the previously incorrect eslint pr, I've tested and created a new PR

yamidark commented:
@apoorva17 Merged the eslint rule PR, and Travis passes now! Do proceed with your PR here.

apoorva17 commented:
@reposense/devs no further changes from my end

damithc commented:
Guys, my apologies. I need to push back on this PR as I realized the multi-color contribution bar is not as visually intuitive as the red contribution bar. With the red bars, it is easy to get a quick sense of relative contribution levels while the multi-color bars need more mental processing. I've asked @apoorva17 to make the breakdown optional. i.e., the normal red bar will appear by default but the user can tick a checkbox in the toolbar to switch to the multi-color one if needed. Hopefully, this addition won't invalidate the work/reviews done already.

apoorva17 commented:
@damithc I've updated the implementation to make breakdown optional as discussed.

apoorva17 commented:
@reposense/devs you may proceed to review the changes made.

damithc commented:
> @damithc I've updated the implementation to make breakdown optional as discussed.

good job on the quick turnaround

apoorva17 commented:
@reposense/devs i would appreciate it if this PR could be merged soon if there aren't any additional changes required as i need the changes in this PR for continuing with #547.

apoorva17 commented:
> The length of the 2 options doesn't seems to match.

@eugenepeh this should be fixed now

apoorva17 commented:
> Hmm... doesn't seems to work well when it is un-grouped too.
> 
> ![image](https://user-images.githubusercontent.com/19277206/52910808-c40c2b00-32d7-11e9-9118-e56405934d5a.png)

What is the problem when it is un-grouped? the length matches default even for un-grouped.

eugenepeh commented:
>What is the problem when it is un-grouped? the length matches default even for un-grouped.

The legend color doesn't match the bar.

apoorva17 commented:
> The legend color doesn't match the bar.

I see, this was because I was generating the legend and assigning colors to filetype by repository, therefore the color for a filetype for one repository was not matching the other. I have now updated the implementation to use a uniform filetype color throughout the report. In the case the it is un-grouped, I have also changed the legend, so that the legend reflects all the filetypes used throughout the report.

apoorva17 commented:
@reposense/devs you may proceed with the review for this PR

yamidark commented on frontend/src/static/js/api.js:
```diff
@@ -58,6 +58,7 @@ window.api = {
             displayName: commits.authorDisplayNameMap[author],
             dailyCommits: commits.authorDailyContributionsMap[author],
             totalCommits: commits.authorFinalContributionMap[author],
+            fileTypeContribution: commits.authorFileTypeContributionMap[author],
```
We are using the term 'file formats' throughout RepoSense, so do refactor the variable and method names to 'fileFormat' as well accordingly (including the backend)

yamidark commented on src/main/java/reposense/authorship/model/AuthorshipSummary.java:
```diff
@@ -1,36 +1,56 @@
 public class AuthorshipSummary {
     private final List<FileResult> fileResults;
     private final HashMap<Author, Integer> authorFinalContributionMap;
+    private final HashMap<Author, LinkedHashMap<String, Integer>> authorFileTypeContributionMap;
```
`HashMap`/`LinkedHashMap` -> `Map`, let's try to use interfaces as much as we can (can change the 1 above as well if you want).

yamidark commented on src/main/java/reposense/authorship/model/AuthorshipSummary.java:
```diff
@@ -1,36 +1,56 @@
+
+        LinkedHashMap<String, Integer> fileTypeContributionMap = authorFileTypeContributionMap.get(author);
+        String fileType = filePath.substring(filePath.lastIndexOf('.') + 1);
+        fileType = (fileType.isEmpty()) ? "others" : fileType;
```
No need for brackets around `(fileType.isEmpty())`

apoorva17 commented:
> Seems like when we group the authors by their repo, the 1st group on top will have all the file formats used, instead of only the ones within his group:
> ![untitled](https://user-images.githubusercontent.com/18352498/53064692-b6d38400-3503-11e9-8b23-1faaeeceb459.png)

As mentioned [here](https://github.com/reposense/RepoSense/pull/487#issuecomment-458016672), when users do not specify any file formats, the legend reflects all the default file formats.

apoorva17 commented:
> Also, as a separate issue, we can probably try to set a minimum width for the contribution bar (and each file format), as some contribution bars are too small to be seen (and to hover over with the mouse to check the number of lines).

Yup, as mentioned [here](https://github.com/reposense/RepoSense/pull/487#issuecomment-457811146), we are thinking of addressing this issue in a separate PR. This change will also need to be applied for the current default contribution bar as the lengths need to match for default and breakdown versions of contribution bar.

eugenepeh commented on frontend/src/static/js/v_summary.js:
```diff
@@ -278,6 +323,24 @@ window.vSummary = {
     },
+    processFileFormats() {
+      const selectedColors = ['#ffe119', '#4363d8', '#3cb44b', '#f58231', '#911eb4', '#46f0f0', '#f032e6',
+        '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1',
```
double indent

apoorva17 replied:
@eugenepeh adding this double indentation seems to be causing CI to fail, i think this eslint rule is not enforced currently?

eugenepeh replied:
@apoorva17 yes most likely the case. Can I trouble you to add in the rule for this as well?
Same as before can be a simple PR, no elaborated message is needed.

apoorva17 replied:
Sure, I've created the PR.

eugenepeh commented on src/main/java/reposense/authorship/model/AuthorshipSummary.java:
```diff
@@ -1,36 +1,57 @@
 
     /**
-     * Increments the contribution count of {@code author} by one.
+     * Increments the contribution count of {@code author} and of a specific fileFormat by one.
```
~and of a specific fileFormat by one.~ `and the corresponding file format specified by {@code filePath} by one.`

eugenepeh commented on src/main/java/reposense/authorship/model/AuthorshipSummary.java:
```diff
@@ -1,36 +1,57 @@
     private final List<FileResult> fileResults;
-    private final HashMap<Author, Integer> authorFinalContributionMap;
+    private final Map<Author, Integer> authorFinalContributionMap;
+    private final Map<Author, Map<String, Integer>> authorFileFormatContributionMap;
```
Since the `LinkedHashMap` is important in this change, better specify it in the variable declaration.
E.g.
`private final Map<Author, LinkedHashMap<String, Integer>> authorFileFormatContributionMap;`

eugenepeh commented on frontend/src/static/js/v_summary.js:
```diff
@@ -145,6 +150,42 @@ window.vSummary = {
                 + `since=${slice.date}'T'00:00:00+08:00&`
                 + `until=${untilDate}'T'23:59:59+08:00`;
     },
+    getFileFormatContributionBars(fileFormatContribution) {
```
this function is a bit long. do add some comments or increase the abstraction.

apoorva17 replied:
I've added in some comments, do let me know if it's ok.

eugenepeh replied:
Yup, its better.

apoorva17 replied:
@eugenepeh There's still an error in Codacy for this: ```Arrow function has a complexity of 5. (complexity)```. Can i increase the complexity? It's difficult to abstract the method because there are quite a few common variables used throughout.

eugenepeh commented on frontend/src/index.jade:
```diff
@@ -155,14 +179,29 @@ html
+                template(v-else)
+                  .summary-chart__contrib(
+                      v-bind:title=
+                          "'Total contribution from ' + minDate + ' to ' + maxDate + ': ' + user.totalCommits\
```
should follow the above style, such that the value of the title does not begin from the next line.

E.g. 
```
v-bind:title="'Total contribution from ' + minDate + ' to ' + maxDate + ': '\
    + user.totalCommits + ' lines'"
```

yamidark commented:
@apoorva17 #578 has been merged and Travis passes now, you may continue with this PR

apoorva17 commented:
@reposense/devs I've made the requested changes and CI passes now, you may proceed with the reviews.

apoorva17 commented:
> Perhaps separate the formats into its own div? it does not compact gracefully on smaller screen size (e.g. with code view opened).
> 
> ![image](https://user-images.githubusercontent.com/19277206/53484682-51f9ca00-3abf-11e9-8f22-e18444461fb7.png)
> ![image](https://user-images.githubusercontent.com/19277206/53484697-5920d800-3abf-11e9-8f2d-3ad097f70c27.png)

@eugenepeh this has been fixed.
@reposense/devs you may proceed with the review of this PR.

yamidark commented:
> > Perhaps separate the formats into its own div? it does not compact gracefully on smaller screen size (e.g. with code view opened).
> > ![image](https://user-images.githubusercontent.com/19277206/53484682-51f9ca00-3abf-11e9-8f22-e18444461fb7.png)
> > ![image](https://user-images.githubusercontent.com/19277206/53484697-5920d800-3abf-11e9-8f2d-3ad097f70c27.png)
> 
> @eugenepeh this has been fixed.
> @reposense/devs you may proceed with the review of this PR.

Tested this out, don't think this is completely fixed? Can make the author name also separate from the formats?
![untitled](https://user-images.githubusercontent.com/18352498/53648810-c4170e00-3c7b-11e9-80b2-376af1bf85ae.png)

apoorva17 commented:
> Tested this out, don't think this is completely fixed? Can make the author name also separate from the formats?
> ![untitled](https://user-images.githubusercontent.com/18352498/53648810-c4170e00-3c7b-11e9-80b2-376af1bf85ae.png)

@yamidark this has been fixed.

apoorva17 commented:
> Can make the text n color marker stay together on same line? Also the scroll bar seems redundant, can it be removed?
> 
> ![image](https://user-images.githubusercontent.com/19277206/53678487-beacd880-3cfa-11e9-91e8-040cf55587b5.png)

Not sure how to keep the 2 elements on same line. Also, I'm not getting the scroll bar on my browser.

eugenepeh commented:
>Not sure how to keep the 2 elements on same line.

You have to prioritize on wrapping the color tag + text pair than it individually. Putting the color tag + text in a single div should solve this problem. Can google on other solutions as well.

>Also, I'm not getting the scroll bar on my browser.

I'm using chrome latest update. It might be the overflow property which cause the appearance of scroll bar or due to font size not fix.

apoorva17 commented:
> Can make the text n color marker stay together on same line? Also the scroll bar seems redundant, can it be removed?
> > ![image](https://user-images.githubusercontent.com/19277206/53678487-beacd880-3cfa-11e9-91e8-040cf55587b5.png)

Ok, this should be fixed now. Not sure about the scrollbar since I cannot view it in the previous version as well.

ongspxm commented:
![image](https://user-images.githubusercontent.com/1430854/53683411-79f56180-3d3b-11e9-988b-418db337d075.png)

apoorva17 commented:
@ongspxm but i've not made any modifications to ```#tabs-wrapper```.

eugenepeh commented:
>Ok, this should be fixed now. Not sure about the scrollbar since I cannot view it in the previous version as well.

Looks good, scroll bar seems to be gone too. Great work 👍

eugenepeh commented:
@yong24s can you help her with the frontend test error?

apoorva17 commented:
@yong24s @ongspxm kindly help me look into the failing frontend test so that we can proceed with this PR.

apoorva17 commented:
@eugenepeh can u assist me with the failing frontend test, as I am not familiar with Cypress and the component giving the error is not something I have modified in this PR? I would like to get this PR merged soon in order to continue with #547.

yong24s commented:
@apoorva17 

It's fine after purging Travis-CI's cache.
I'll investigate.

apoorva17 commented:
@reposense/devs no further changes from my end, you may proceed with this PR.

apoorva17 commented:
@reposense/devs you may kindly proceed with the review for this PR.

ongspxm commented on frontend/src/static/js/v_authorship.js:
```diff
@@ -257,6 +248,13 @@ window.vAuthorship = {
       return this.files.filter((file) => this.isSelected(file.path)
           && minimatch(file.path, this.filterSearch, { matchBase: true }));
     },
+    filesExistingLinesObj() {
```
minor nit, follow existing name format? `getExistingLinesObj()`? any relevant name

apoorva17 replied:
sure, i've changed from ```filesExistingLinesObj()``` to ```getExistingLinesObj()```

ongspxm commented on frontend/src/static/js/v_summary.js:
```diff
@@ -281,6 +329,24 @@ window.vSummary = {
+    processFileFormats() {
+      const selectedColors = ['#ffe119', '#4363d8', '#3cb44b', '#f58231', '#911eb4', '#46f0f0', '#f032e6',
+          '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1',
+          '#000075', '#808080'];
```
may want to expose the mui-colors and make those into classes and use them instead. this one seems quite hacky. but can leave for another time. the exposing of mui-colors, we can create a new issue for that if needed.

apoorva17 replied:
wouldn't it be better to use an object instead of class in js? i think it would be good to create a separate issue for exposing the mui-colors, this will allow discussion on how the colors can be incorporated into other potential uses as well.

apoorva17 commented:
@reposense/devs requested changes have been made.

apoorva17 commented:
@reposense/devs do you require any other changes? If not, I would appreciate it if this PR could be merged soon so that I can proceed with #547

yamidark commented:
@apoorva17 

Seems like your PR proposed merge message does not contain the `{why it needs to change}` section. Do add this in (the 2nd paragraph).
Also, as mentioned in our [commit organization](https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-organization), do avoid the use of words like 'currently', 'originally' in your message, as it is already implied.

apoorva17 commented:
> @apoorva17
> 
> Seems like your PR proposed merge message does not contain the `{why it needs to change}` section. Do add this in (the 2nd paragraph).
> Also, as mentioned in our [commit organization](https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-organization), do avoid the use of words like 'currently', 'originally' in your message, as it is already implied.

@yamidark i've updated the merge message.

yamidark commented:
@apoorva17 Do take note that each line should be at most 72 characters only. Also, you still seem to be mixing the term `filetype` and `file formats` in your message. Do follow the terms and conventions that this project is using.

I have edited your merge message accordingly, do take note of these in the future.

apoorva17 commented:
> @apoorva17 Do take note that each line should be at most 72 characters only. Also, you still seem to be mixing the term `filetype` and `file formats` in your message. Do follow the terms and conventions that this project is using.
> 
> I have edited your merge message accordingly, do take note of these in the future.

Ok, thank you.
